### DISCUSSION TITLE
PyScript Next

### DISCUSSION THREAD

**Discussion by WebReflection** [pyscript/pyscript#1569] [General]
We would like to discuss as a team what we need to do to get *PyScript Next* out of the current TP.

This discussion goal is to have some sort of "_epic_" so that we can both discuss and track all points (current written in priority order).

- [ ] **Automated Tests** - we are currently testing manually multiple interpreters and other parts of the project work as expected. This easily leads to possibly missing tests before further releases and current manual test also doesn't help spotting possible regressions out of CI jobs, or locally. We already have integration tests for "*classic PyScript*" but things with core and the new custom *py* type are so different that it's hard to estimate what can be reused VS a new/better approach focused around both multiple interpreters and the ability to have workers + workers utilities.
- [x] **[Better Debugging](https://github.com/pyscript/pyscript/pull/1572)** - there is a `NO_MIN=1 npm run build` option to build non minified code but the DX is still problematic due the fact we inevitably embed the *Worker* code as *Blob* (that is a *JS* string). I believe it would be great if we could have an `npm run dev` that orchestrate a few things differently, such as: 1. it uses non minified code 2. it uses a temp file instead of a blob to simplify debugging code that lands in *workers* too 3. as stretch goal, it might generate a JSON source map to copy/paste in the tested page so that it's easier to live code and have that code reflected, as opposite of needing to build over and over 4. as extra stretch goal of point 3, we could have a watcher to the `esm/` folder and re-build only such temporary file with the new *worker* content ... if all of these would be in place, debugging would be a way more pleasant experience for both ourselves and possible contributors.
- [ ] **PyScript features** - beside simplifying integration with `<py-config>` and aligning with classic `<script type="py">` and the `<py-script>` custom element counter part, it's not clear what we should offer as the first-class custom type citizen of this project. I believe this discussion or list of features such as `<py-repl>` or `<py-terminal>` or whatever extra feature we'd like to add should be an *epic* a part but this topic can't be omitted in this discussion.
- [ ] **Better error reporting**: https://github.com/pyscript/pyscript/issues/1596
- [ ] **Port examples**: in addition to moving the examples out of the main repo, we agree that current examples should work (or adabpted to work) with Next (in that sense, Next should be as close as possible to feature parity with the latest release)

---

## Next.next()

- [ ] **Custom types plugins** - there are tons of hooks exposes to define a custom *type* and all hooks can already be used to define each type custom *plugins* (i.e. `<py-repl>` or other `py` custom type related plugins). The current API though just exposes hooks but it doesn't offer much more than that ... should we improve this story?
- [ ] **Repl** - < py-repl > will be changed to actually look like the "usual" Python interactive prompt, via something like [Xterm.js](https://xtermjs.org/) 👍 
- [ ] **`<py-notebook>`**: The current `<py-repl>` (that is not a real REPL !) will be renamed to `<py-notebook>`, which will contain `<py-cell>` tags. 👍 --> Details to be discussed on its own issue or a discovery PR
- [ ] **Python Plugins**: Python plugins are specific instances of "runtime plugins" (e.g. in Lua, Ruby etc). As such, the needed hooks to create plugins should already be there. In tha sense we need to: 1. have a documented example (yet). -> 2. add docs  3. evaluate if actually all hooks are there. [4.  open a issues for the additional APIs/features missing]
- [ ] **MicroPyScript** - people just want to use *MicroPython* out there, and the sooner we also get in charge of the **mpy** custom type, the better it is for everyone, so that we can focus on providing the best *PyScript* experience out of *MicroPython* interpreter.

---

- - -

Out of Scope

- - -

* A/Sync only interpreters
* Other Interpreters
* Other Config

- - -

To Be Defined

- - -

- [ ] **A/Sync only interpreters** - we have some interpreter incapable of offering both *sync* and *async* APIs to evaluate and execute code. This might lead to a lot of confusion when **webR** as example is used, but its `run(code)` wrap helper can't possibly work synchronously as the project does not expose (yet?) a way to deal with it. The interoperability among interpreters for similar tasks (yet specialized per interpreter) might suffer but we need to decide if we really want to allow/enable interpreters that don't offer both APIs, being the **sync** one actually the most interesting one for general purpose use cases.
- [ ] **Other Interpreters** - coupled with the previous point, we should try to understand if the status of these interpreters would actually benefit *core* or its users, as opposite of creating just frustration. As example, *wasmoon* has been announced as experimental but there's pretty much nothing it cannot do compared to *pyodide* or *micropython* **but** the *ruby-wasm-wasi* interpreter can't really do much at its current state, and we don't get to plan 3rd party software releases plans or direction (we can of course contribute there though). This point is just to be sure if we claim that *WebR* will be next as supported interpreter, we need to carefully explain what are the limitations, and that's true for any other interpreter we'd like to have ... we need investigation over capabilities before stating it will be in.
- [ ] **Other Config** - all interpreters are currently using the *config* format inherited from "*classic PyScript*" but while I think that config is generic enough to welcome any other interpreter config needs, it was surely strongly coupled with what *pyodide* only could do (i.e. *micropip* only dedicated section for runtime imports that might never be possible in other interpreters). Are we sure that's still the format we'd like to support forward? It's great that part has also a good documentation, so I am for the "*KISS*" approach here but maybe it's worth considering something else?

**Comment by WebReflection** [comment]
From today sync:

### Automated Tests

We all agree we want integration tests out of the box, both locally and CI. We also agreed that performance is not the top priority while having ported some orchestration from "*pyscript classic*" would be nice, as long as anything strongly coupled with *pyodide* only runtime goes away.

### Better Debugging

We agreed that this is not just helpful for ourselves or contributors, we want to be able to hook into better debugging possibilities offered by any interpreter (if possible).

### PyScript features

Too little time to talk about this, we're synchronizing again tomorrow around this specific topic.

### MicroPyScript

It looks like we need to solve *PyScript features* before we can even decide what should be in. No particular friction in agreeing this might be indeed a super welcomed feature to have (`<script type="mpy">` and `<mpy-script>` with all the good things we believe `<py-script>` could bring in in a similar fashion to MicroPython).

### Custom types plugins (PyScript plugins)

Hopefully a glance of what we want/need/do after tomorrow sync.


### Other topics

No time to refine them, but these are indeed lower priority so far ... we're good.

**Comment by fpliger** [comment]
Thanks for the summary! In order to support the conversation about Next PyScript Features, I'll list a few things that should be on the agenda (not developing the topics or giving an opinion on what should happen on purpose. ;) ):

In PyScript Classic:

* py-repl
* py-terminal
* splashscreen
* Python Plugins

Different Behaviour or worth discussing due to new features made possible by the new core

* `<py-script>` behavior when managing the last element in a script
* `display` (I'm not entirely sure here.. I may be wrong that they differ)
* multiple configs per interpreter vs single config vs both 
* Default `<py-script>`/`<script type=xxx>` behavior (main thread? worker? sync? async?)
* Workers interface
* low level "plugins" 
* PyScript higher level [Python] API (i.e. `display`, etc..)

Please feel free to add more.

**Reply by antocuni** [reply]
More topics which I would like to talk about:
- Error reporting: how we want to report errors/warnings to the user, and whether it should be responsibility of core or anything on top. IMHO this should be very high in the priority list

- "global" filesystem: we are going towards a model in which every "instance of interpreter" is basically equivalent to a process. AFAIK, currently each of them has its own independent FS, provided by emscripten (at least for the interpreters based on it). We should think carefully about what is the model that we want, because currently if you write to a file in lua and try to read it from python, it doesn't work. (Not saying that it should, but we should think about this). Probably not top priority, though.

**Reply by antocuni** [reply]
- event handlers: this is a can of worms, I know. The current state is `<button py-click="foo()">`, where the content is a string which is interpreted/evaluated on the fly. The alternative approach is to use a callable instead (and the callable should receive the `event` argument). I know that we have discussed this a lot in the past, but I have an additional data point now: what about compiled languages, i.e. those which produce a .wasm file? If the content of event handlers is a callable, it is straightforward. If it's a string to exec, much less.

**Reply by WebReflection** [reply]
`<button py-click="foo(event)">` already works though, not sure I am following

**Reply by WebReflection** [reply]
> what about compiled languages, i.e. those which produce a .wasm file?

we don't have any right now so this is something to discuss **when** we'll have that issue but it's clear to me anything that lands and export functionalities must have those functionalities available ... what's a callable in an inline listener defined in behavior by standards? if *nothing* is the answer, then declarative `xxx-click` that works as standards is not the place to fix that.

**Reply by WebReflection** [reply]
if the *callable* is a name, then `callable(event)` can be parsed and served as `callable` + `event` ... so I don't think we should destroy 20 years of HTML declarative inline behavior for the sake of callables there, but again, I don't know what is a callable in here and why that's any different from the current state in practice.

**Reply by antocuni** [reply]
> `<button py-click="foo(event)">` already works though, not sure I am following

imagine that we want to support Rust or AssemblyScript. In that case, we won't have the source code to interpret at runtime, but instead the user would compile the code offline and ship a compiled .wasm file.
This .wasm file would contain a set of "exported functions", which can be called from JS -- including the event handler functions.

No, if pyscript.core see a `rust-click="foo(event)"`, what does it do? In order to understand what is written between the quotes, you need a rust compiler, but you don't have it at runtime.

That's why I'm saying that we *might* want to consider a different approach (not saying that we should do it, but at least discuss). If we have `rust-click="foo"`, then it's very easy for pyscript.core to look up the function `foo` from the exported functions and call it.

**Reply by WebReflection** [reply]
the `rust-click="foo"` will receive the event so it's not different at all to `rust-click="foo(event)"` which is how inline / declarative listeners always worked to date. We had that before and it got nuked because `@when` decorator solved that in *Python* world + all it takes to address these speculative cases is to add a helper that invokes that function with the current event ... branching syntax for the sake of it seems confusing and unnecessary. If these PLs have just exported names, they can only take an `event` as argument and I think there's no syntax that couldn't be easily translated to that but consistency for documentation and Web friendliness will remain intact.

I don't like too much to talk about non necessary, or not even implemented or available at all, stuff when there is so much else necessary now to talk about but I also don't think I'll personally change my mind around this topic and having branched logic for inline listeners looks for troubles to me, it doesn't solve much (imho).

In short, I want to avoid features creep for hypothetical non-currently-existing use cases.

**Reply by WebReflection** [reply]
P.S. worth saying that we can easily implement `xxx-click="foo"` as alternative so it's not that I am against it, I just don't like too much branching out expectations, behaviors, or syntax, for use cases we don't need to tackle today. If we all want to have that, we can do it, but somebody gotta write the documentation too for it ;-)

**Reply by WebReflection** [reply]
@antocuni let's talk hypothetical though ... if I had `pyodide-click="foo"` how would I avoid using `runPython` there at the end of the day? Is there any `pyodide.runMethod("foo", event)` there? (maybe I can just check the API) ... that's to say that if there's no mechanism to make it possible to use same syntax across interpreters, I might be skeptical this would be more useful than *us* finding some meaningful helper for Rust or AssemblyScript instead of making all other interpreters more problematic to deal with ... but maybe that's not the case.

**Reply by WebReflection** [reply]
@antocuni OK, this was actually easy-peasy to me to both implement and test so I hope we can move forward around this topic: https://github.com/pyscript/pyscript/pull/1574

The benefits are not only for "future Rust or AssemblyScript", the benefits are kinda relevant already for currently available interpreters too, as there are less moving parts, and (I believe) some sort of performance boost too.

**edit** *TL;DR* I don't necessarily like it but I see benefits around this choice and after all we're doing stuff nobody on the Web would've thought about before :D ... I'm sold this is a good, simple, thing to offer.

**Reply by WebReflection** [reply]
> If we have rust-click="foo", then it's very easy for pyscript.core to look up the function foo from the exported functions and call it.

The PR addresses exactly that. If you're not happy about that PR please comment in there.

If you have more to say to what you proposed yourself, either write that in the PR or here.

The PR is there because your concerns for the future of the project and possible interpreters might be valid and, coincidentally, reduces the amount of moving code when a function is exposed, simplifying the `event` dance too.

We can have the best options from all worlds because there's no disambiguation issue in what you suggested, but I am afraid having *only* your suggestion shouldn't happen (imho), as that breaks everyone expectations from the Web and previous PyScript capabilities.

I hope the dual, never ambiguous, compromise, works well for us today, and for your concern tomorrow.

**Reply by ntoll** [reply]
More thoughts to follow up on @fpliger's list - and in light of the video call catchup discussion yesterday.

When it comes to PyScript classic tags, in yesterday's call we discussed them and came up with the following:

* `<py-script>` - as it currently behaves.
* Rename `<py-repl>` (it's not a REPL) to `<py-notebook>` containing `<py-cell>` tags.
* Have a `<py-repl>` that looks like the "usual" Python interactive prompt, via something like [Xterm.js](https://xtermjs.org/)
* Python plugins are specific instances of "runtime plugins" (e.g. in Lua, Ruby etc). As such there's already the needed hooks and so on to be able to create such plugins... we just don't have a documented example (yet).

We didn't discuss the splash screen. Personally, I don't think we need it since it adds complexity, and the user (who is writing the code) can hook into existing events to add/remove their own if they want/need it. We should just document how to do it (as per runtime based plugins).

**Reply by ntoll** [reply]
Regarding @antocuni's list:

* Error reporting... looks like @WebReflection's most recent "better debugging" MR addresses much of this. Not quite the same as error reporting, but if we re-implement the `<py-repl>` tag as we discussed in yesterday's meeting, and ran Pyodide from a worker, we should be able to put together a relatively simple example of how to use `pdb` in PyScript. This is another example of a class of problem that I'm naming `not-yet-documented` - i.e. a feature is already potentially there, we just need to highlight *how* to make it work.
* Global filesystem... I like the fact that everything is isolated. You'll be totally unsurprised to hear, @antocuni, that I like the idea of message passing between workers/instances for the sharing of data. The browser does this very efficiently already, since that's how JavaScript works... so I feel this is a case of "when in Rome..." 🇮🇹 It also makes it very easy to explain: "Your filesystem is isolated to your interpreter, but if you want to share data, use this standards based method which we've documented over here..."
* Event handlers... I've always felt uncomfortable that we basically `eval` the handler code. I also don't want to wade into a technical discussion that's already progressing and cause the waters to muddy. Rather, if I understand @WebReflection's suggestion correctly (specify the function name, all such functions receive the JS `event`), then this is both avoiding eval-ing arbitrary code, complements how the browser currently works and, importantly, is easy to explain.

**Comment by WebReflection** [comment]
**Update** - [Better Debugging MR landed](https://github.com/pyscript/pyscript/pull/1572)

  * it solves all points except the sourcemap (we can have a MR a part for that if we want/need it)
  * it does not solve better debugging for each interpreter (we can have a MR a part for that if we have a way to enable that)
  * it got escalated because for integration tests, having builds with non optimized code might be desirable, so it's necessary to bring in the best possible **Automated Tests** story

**Comment by WebReflection** [comment]
**Update** - Automated Tests

  * I've landed after some headache *playwright* for [coincident](https://github.com/WebReflection/coincident/pull/14)
  * there are **tons of timing issues** there due:
    * the need for a server able to provide all required headers or nothing works ... I didn't spend too much time in there with a `spawn` instead of an `exec` with listeners on *stdout* instead of a `setTimeout(ready, 500)` ... but both "*my machine*" and *CI* seem to be happy about that `500` (and yes, I hate that ... I might do soon a better job there ... )
    * the need for a listener about any created *Worker* on the page ... I am not even going down the "*multiple workers*" rabbit hole for now, but of course I could (and likely will at some point)
    * the limitation around the *Worker* listener ... it offers only a `close` event and nothing else to signal it's parsed and executed
    * the limitation around the fact previous point is hard, so I need naive timers to assume previous point happened
  * because I am faster with *JS*, I just use *Playwright* for node ... we could revisit that in PyScript but after all the important things of integration tests is within dedicated files and each interpreter these use ... not about how the test is written itself, so maybe we can stick with that?
  * all dances before tests are different from current PyScript so that I don't think there's much I can reuse from "*classic PyScript*" tests but I'll definitively port meaningful tests over

**Reply by WebReflection** [reply]
P.S. everything we have now does not work well with workers, we can't define workers within the tests (unless we orchestrate Blobs all over the place which might be awesome but I've no idea if either possible or how long it'd take) so the expectations will remain in tests but I think the logic should be in real files out there, dedicated to satisfy expectations. That's the fun part of multiple interpreters, we could simulate Pyodide same way we did before, but everything else will be different and ... if it's integration we're after, specially for workers we can't use inline code within the test ... we need to be structured and clean in how we approach this (imho).

**Reply by WebReflection** [reply]
P.S.2 the bootstrap time is relatively easy to solve and indeed now I have solved it in *coincident* ... the Workers timing though is still something I've no idea how to solve (yet)

**Comment by fpliger** [comment]
DISCUSSION UPDATES: (notes taken during meeting between @WebReflection @ntoll @antocuni and myself

----

In the topic of debugging. There are 3 missing features relating to that:

1. Debugging for the end users
2. Debugging for contributors (that is done)
3. Error reporting

We need to open issues for 1 and 3. @antocuni will take care of opening 1 and I will take care of 3

-----

Topics that are out of the scope of the first version of the next:

* A/Sync only interpreters
* Other Interpreters
* Other Config

-----

* **< py-script >** - as it currently behaves. 👍 
* **Rename < py-repl >** (it's not a REPL) to < py-notebook > containing < py-cell > tags. 👍  --> Details to be discussed on its own issues
* **Have a < py-repl > that looks like the "usual" Python interactive prompt**, via something like [Xterm.js](https://xtermjs.org/)  👍  
* **Python plugins are specific instances of "runtime plugins"** (e.g. in Lua, Ruby etc). As such there's already the needed hooks and so on to be able to create such plugins... we just don't have a documented example (yet).  -> 👋  (italian hand waving) First step of this is to add docs and evaluate if actually all hooks are there. If so, 👍 , otherwise we open an issue for the additional aspects missing

-----

Last topic discussed: **PyScript features**

The decision here, in the scope of the first release of Next, is to have parity for all things that are not explicitly called out (for instance, `py-repl`)

**Comment by fpliger** [comment]
@WebReflection @ntoll @antocuni @JeffersGlass  Heads up. I'm adding some of the items we agreed on to the description on top so that we can have the checklist in the description.  I'll also link these to issues/PRs related to them so that we can better track things.

**Reply by antocuni** [reply]
I have opened an issue about [better error reporting](https://github.com/pyscript/pyscript/issues/1596) and added it to the bulleted list above

**Reply by antocuni** [reply]
and here is a discussion containing my thoughts about python debugging: https://github.com/pyscript/pyscript/discussions/1597

**Reply by fpliger** [reply]
Thank you! Yesterday I've also started to look into porting some of the more complex examples (Panel examples for instance) and added an extra item above

**Comment by antocuni** [comment]
@fpliger I just noticed this item in the TODO list above:
> Repl - < py-repl > will be changed to actually look like the "usual" Python interactive prompt, via something like [Xterm.js](https://xtermjs.org/) +1

I think it should be slightly lower priority than the rest. It's a new feature compared to pyscript-classic, and it requires a lot of thinking/planning to understand what is the best way to implement it in such a way that works in both the main thread and worker cases.

**Reply by fpliger** [reply]
Agree.. Actually, the same is valid for `< py-notebook >`.  

Honestly, the approach that @JeffersGlass is taking in #1585 seems in the right direction (PR to start with a PoC where we can discuss over an initial idea... :)

**Comment by fpliger** [comment]
@WebReflection @antocuni @JeffersGlass  based on the recent discussions and all the things happening at the same time, I think we may have defined too much as part of Next and would benefit from splitting it and releasing things in 2 steps:

1. Next Release of PyScript with most of what we mentioned above but keeping things as close to feature parity as possible but based on polyscript. When I say as close to feature parity I don't mean trying to force old odd behaviors into the new architecture based on Polyscript, if there's a better way let's fast-forward old things into how they should have been. I mostly mean that we should not try to add too many features "on purpose" and add too much to chew into the release with the new core. Keeping it mostly a core change (again, with the changes needed and not much more), is beneficial to understand if we didn't break things in the meantime and would allow users an easier/lighter transition
2. We iterate on 1, add fixes or anything learned from 1 but, more that, we iterate on the features we want to change (i.e., repl, py-notebook, improved debugging, etc..)

For this, I think the first thing we should do asap is to fully replace the current core to use polyscript. From there, we can make faster progress on the items and features we defined above in this thread

**Reply by WebReflection** [reply]
The way core bootstraps custom types is different and I already tried to use core in classic PyScript and failed because the work behind Synclink made it too hard to just swap core ... everything is too stronly coupled with a lot of code not needed due Synclink.

There is already a PyScript Next based on polyscript and bringing what we want there is way faster and less error prone than just swapping core with polyscript so I think we should keep doing that.

**edit** the current PyScript Next is [here](https://github.com/pyscript/polyscript/blob/main/esm/custom/pyscript.js) and that should be moved into PyScript with polyscript as its core dependency.

**Reply by WebReflection** [reply]
Note a part: @antocuni expressed the idea/desire of having pyodide always running from a *Worker* ... that means all scripts *py* or *py-script* content should be forwarded and executed in such worker but also the bootstrap would be very different ... do we agree this is the way forward we want for PyScript next? I think it might be risky due amount of module that could possibly assume `js` exists and so does its `document` but as discussed with @fpliger we could also circumvent that and provide our own `document` in the Worker env ... what are your thoughts here? I think it's doable but if that's the road ahead that's the most core thing for PyScript Next we should work on.

@JeffersGlass your opinion also appreciated here.

**Reply by fpliger** [reply]
I didn't necessarily mean that we should keep the whole codebase and just swap core but was trying to see it at a higher level, in terms of features. For instance, you mentioned Antonio's preference to run scrips in a worker by default, in that case we should not change how it works in the next release even if we all (hypothetically) agreed it should be the default. We should wait at  least an iteration to "do the right thing" and should do what we do today (or something very close) first.

**Reply by WebReflection** [reply]
+1 to that here

**Reply by fpliger** [reply]
Also, I was implicit about it but think we should build from the next branch or just start a new one clean that uses polyscript

**Reply by JeffersGlass** [reply]
+1 from me as well to the next PyScript release not adding too many features "on purpose". Personally, I'm not sure how I feel about PyScript _always_ running in a worker, but either way that does feel like something to push to a future release. Ditto the new py-repl/py-notebook ideas.

**Reply by antocuni** [reply]
+1 on everything which was written above.

To clarify my idea about workers. I think that *eventually*, in the long run, we want the following:
1. it should be possible to run python code in the main thread
2. it should be possible to run python code in the worker (in the most transparent way that it's realistically possible)
3. user should be able to decide whether to use (1) or (2)
4. I *think* that `worker` should be the default (but we can debate about that, of course).

But this doesn't mean that all of the above should happen in the next release: I am perfectly fine to introduce changes with a more incremental approach. For example, a tentative roadmap could be:
1. in release N we have "feature parity" but based on polyscript
2. in release N+1 we introduce the option to run code in a worker, but it must be opt-in
3. *if* we see that workers are working well, in release N+2 we make them the default

Note that for example, python-in-the-worker is a prerequisite for python-level debugging

**Reply by WebReflection** [reply]
as so far everything was on main (and most stuff we use assume that via `js.document` access for Pyodide) it would be really breaking to change default execution but more importantly, the subtle yet important differences between worker code and main code might just break even stuff that already works ... so I'd say we should consider a `thread = main | worker` *config* option to allow the default definition and/or think about a `thread="main | worker"` attribute that can disambiguate.

Then again, we're talking about many *if* ... *if the worker story is super successful* and *if we manage to talk again about this for N+2* :sweat_smile:
